### PR TITLE
Migrate to zlib-ng, part 3: Remove zlib and zlib-intel source code and license mentions

### PR DESCRIPTION
**Pull request by carlossanlop**
Contributes to: https://github.com/dotnet/runtime/issues/101465

Split in commits for easier review.

### CODE REVIEW

**Comment by jkotas** [comment]
You should be also able to delete zlib-dev package from the Linux build prereqs at https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md#toolchain-setup

**Comment by jkotas** [comment]
And from https://github.com/dotnet/runtime/blob/main/eng/install-native-dependencies.sh#L27-L31

**Comment by jkotas** [comment]
And from under https://github.com/dotnet/dotnet-buildtools-prereqs-docker/tree/main/src/azurelinux/3.0/net9.0

**Review by jkotas — APPROVED** [review]

**Comment by carlossanlop** [comment]
> You should be also able to delete zlib-dev package from the Linux build prereqs at https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md#toolchain-setup

Questions:
- Should we change this doc to also explain that system zlib is used in armv6, android, ios, tvos and maccatalyst?
- Is system zlib required for a cross build?

**Comment by jkotas** [comment]
> Should we change this doc to also explain that system zlib is used in armv6, android, ios, tvos and maccatalyst?

From the doc: "These instructions are written assuming the current Ubuntu LTS.". armv6, android, ios, tvos and maccatalyst concerns are not relevant for this doc.

> Is system zlib required for a cross build?

Yes - if the target platform is configured to system zlib. I cannot think about any docs where it would need to be mentioned.

**Comment by carlossanlop** [comment]
> I cannot think about any docs where it would need to be mentioned.

I found a android coreclr cross building, but couldn't find one for the apple mobile platforms: https://github.com/dotnet/runtime/blob/898ffc5f74fad1593ed4036a69da039937224f99/docs/workflow/building/coreclr/android.md

I also found a general one for coreclr cross building: https://github.com/dotnet/runtime/blob/898ffc5f74fad1593ed4036a69da039937224f99/docs/workflow/building/coreclr/cross-building.md

The mono build document is very generic but has links to android and ios testing (not cross building): https://github.com/dotnet/runtime/tree/898ffc5f74fad1593ed4036a69da039937224f99/docs/workflow/building/mono

Do any of those look good for mentioning the system zlib requirement?

**Comment by carlossanlop** [comment]
@jkotas Oh, the first doc you shared has a cross build section a few lines below, we could move `zlib1g-dev` there: https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md#additional-requirements-for-cross-building

**Comment by carlossanlop** [comment]
> And from under https://github.com/dotnet/dotnet-buildtools-prereqs-docker/tree/main/src/azurelinux/3.0/net9.0

OK I see that azurelinux is the only distro we split depending on the .NET 9.0 version. I mention this because I see zlib-devel and zlib1g-dev being installed in most of the other Dockerfiles but they should remain untouched as these images are shared by all the servicing branches.

File: docs/workflow/building/libraries/README.md
Comment (jkotas) [inline docs/workflow/building/libraries/README.md:120]:
```suggestion
The libraries build contains some native code. This includes shims over libc, openssl, gssapi, and zlib. The build system uses CMake to generate Makefiles using clang. The build also uses git for generating some version information.
```
I think it is more appropriate to just say zlib here. The shim can be for system zlib or zlib-ng.

Similarly, we just say libc here even though it can be glibc, musl-libc or bionic-libc.

File: docs/workflow/building/libraries/webassembly-instructions.md
Comment (jkotas) [inline docs/workflow/building/libraries/webassembly-instructions.md:68]:
```suggestion
The libraries build contains some native code. This includes shims over libc, openssl, gssapi, and zlib. The build system uses CMake to generate Makefiles using clang. The build also uses git for generating some version information.
```

File: src/libraries/System.IO.Compression/README.md
Comment (jkotas) [inline src/libraries/System.IO.Compression/README.md:7]:
```suggestion
[../../native/libs/System.IO.Compression.Native](../../native/libs/System.IO.Compression.Native) contains the PAL (Platform Abstraction Layer) of the zlib library used by System.IO.Compression.
```

**Review by jkotas — APPROVED** [review]

**Comment by carlossanlop** [comment]
Can't merge this yet because part 2 was just reverted: https://github.com/dotnet/runtime/pull/104414

I'm working on a fix for it and will come back to this PR after the fix is merged.

**Comment by carlossanlop** [comment]
@jkotas I removed the `no merge` label as it seems we're in a good state.
If there are no objections, I'll merge this today (the CI is restarting because of a merge conflict with an *.md file).

File: docs/workflow/requirements/linux-requirements.md
Comment (jkotas) [inline docs/workflow/requirements/linux-requirements.md:60]: Could you please cleanup the additional zlib references that were added in the meantime?

https://github.com/dotnet/runtime/pull/104860#discussion_r1677037992

File: eng/install-native-dependencies.sh
Comment (jkotas) [inline eng/install-native-dependencies.sh:31]:
```suggestion
            dnf install -y cmake llvm lld lldb clang python curl libicu-devel openssl-devel krb5-devel lttng-ust-devel
```

Comment (carlossanlop) [reply eng/install-native-dependencies.sh:31]: My bad, I resolved the merge conflicts incorrectly. Thanks for catching this.

File: docs/workflow/requirements/linux-requirements.md
Comment (carlossanlop) [reply docs/workflow/requirements/linux-requirements.md:60]: Fixed. Removed the extra ones from linux-requirements.md and install-native-dependencies.md.

**Comment by carlossanlop** [comment]
This is removing dead code + adding some markdown changes. The previous CI runs passed (the dead code removal does not affect the build), and the latest comment simply fixed a couple of markdown files. I'll fast merge this.

**Comment by carlossanlop** [comment]
In the previous run there was only one failure that did not get linked to a KnownBuildIssue because an issue did not exist at the moment:
- https://github.com/dotnet/runtime/issues/105176
  - https://dev.azure.com/dnceng-public/public/_build/results?buildId=748332&view=logs&j=2eda3965-9146-55d1-4226-c3d3f8432d8f&t=27bbe0a6-6c05-550a-bc6f-305cf5eac8aa&l=117 
-

**Comment by jakobbotsch** [comment]
@carlossanlop After this PR we are getting build failures on a clean ubuntu-22 image:
```
  /usr/bin/ld.bfd: cannot find -lz: No such file or directory
clang : error : linker command failed with exit code 1 (use -v to see invocation) 
```
```
...libSystem.Security.Cryptography.Native.OpenSsl.a -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -Wl,-e0x0 -pthread -ldl -lz -lrt -lm -shared -Wl,-z,relro -Wl,-z,now -Wl,--eh-frame-hdr -Wl,--discard-all -Wl,--gc-sections" exited with code 1.
```
This is when running `eng/install-native-dependencies.sh && ./build.sh ...` (full script is https://gist.github.com/EgorBo/0e961d7fdf9a763fe4f78cf502115f29).
It succeeds if `zlib1g-dev` is manually installed.

**Comment by carlossanlop** [comment]
Okay, I'm reverting it to avoid blocking today's snap.

**Comment by carlossanlop** [comment]
@jakobbotsch reverted, sorry about the noise.

**Comment by ericstj** [comment]
Is this another case where folks are using scripts that aren't represented in our actual build process?  Or is it a public build vs CI validation thing?

**Comment by EgorBo** [comment]
> Is this another case where folks are using scripts that aren't represented in our actual build process? Or is it a public build vs CI validation thing?

It's probably not represented, yes, but it follows the official build doc. Basically, a clean new Azure VM with Ubuntu x64
```bash
eng/./install-native-dependencies.sh
./build.sh Clr+Libs -c Release
```
fails with the error posted above ^, after I added a manual `apt install -y zlib1g-dev` it has fixed the issue

**Comment by jkotas** [comment]
We would run into the same build break in the CI and official builds once https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/1125 is merged and propagates through the system.

> Is this another case where folks are using scripts that aren't represented in our actual build process?

We have two documented Linux build instructions: https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/linux-instructions.md:
- Containers that are used and validated in CI and official builds.
- "Local" build environment that just requires installing several packages. It is more productive setup for development. It is not validated by dotnet/runtime CI. It is sort of validated by source build.

**Pull request merged**
